catalog/service/rest/src/main/java/org/projectnessie/catalog/service/rest/IcebergErrorMapper.java

```diff
@@ -55,29 +57,36 @@ public class IcebergErrorMapper {
   private static final Logger LOGGER = LoggerFactory.getLogger(IcebergErrorMapper.class);
 
   private final ExceptionConfig exceptionConfig;
+  private final BackendExceptionMapper backendExceptionMapper;
 
   @Inject
-  public IcebergErrorMapper(ExceptionConfig exceptionConfig) {
+  public IcebergErrorMapper(
+      ExceptionConfig exceptionConfig, BackendExceptionMapper backendExceptionMapper) {
     this.exceptionConfig = exceptionConfig;
+    this.backendExceptionMapper = backendExceptionMapper;
   }
 
   public Response toResponse(Throwable ex, IcebergEntityKind kind) {
-    if (ex instanceof CompletionException) {
-      ex = ex.getCause();
-    }
-
-    if (ex.getClass() == RuntimeException.class && ex.getCause() != null) {
-      ex = ex.getCause();
+    IcebergErrorResponse body = null;
+    for (Throwable th = ex; th != null; th = th.getCause()) {
```

Comment:
In case of a backend exception we traverse the cause-chain twice: here and in `BackendExceptionMapper.analyze()`.

WDYT of having this in `BackendExceptionMapper`:
```java

  public Optional<BackendErrorStatus> analyze(Throwable ex) {
    for (Throwable th = ex; th != null; th = th.getCause()) {
      Optional<BackendErrorStatus> status = analyzeNoChain(th);
      if (status.isPresent()) {
        return status;
      }
    }

    return Optional.empty();
  }

  public Optional<BackendErrorStatus> analyzeNoChain(Throwable ex) {
    for (Analyzer analyzer : analyzers()) {
      BackendErrorStatus status = analyzer.analyze(ex);
      if (status != null) {
        return Optional.of(status);
      }
    }

    return Optional.empty();
  }
```
and changing this function to
```java
  public Response toResponse(Throwable ex, IcebergEntityKind kind) {
    IcebergErrorResponse body = null;
    for (Throwable th = ex; th != null; th = th.getCause()) {
      if (th instanceof BaseNessieClientServerException) {
        BaseNessieClientServerException e = (BaseNessieClientServerException) th;
        body = mapNessieError(e, e.getErrorCode(), e.getErrorDetails(), kind);
        break;
      } else if (th instanceof IllegalArgumentException) {
        body = errorResponse(400, "IllegalArgumentException", th.getMessage(), th);
        break;
      } else if (th instanceof IcebergException) {
        body = ((IcebergException) th).toErrorResponse();
        break;
      }
      Optional<BackendErrorStatus> status = backendExceptionMapper.analyzeNoChain(ex);
      if (status.isPresent()) {
        body = mapStorageFailure(status.get(), ex);
        break;
      }
    }

    if (body == null) {
      LOGGER.warn("Unhandled exception returned as HTTP/500: {}", ex, ex);
      body = errorResponse(500, ex.getClass().getSimpleName(), ex.getMessage(), ex);
    }

    Integer code = body.error().code();
    return Response.status(code == null ? 500 : code).entity(body).build();
  }
```

Reply:
Yes, I was thinking about that too... will do.

Reply:
Implemented... slightly differently :)